bootstore/src/schemes/v0/state.rs

```diff
@@ -0,0 +1,594 @@
+    Timeout,
+}
+
+impl RackInitState {
+    // Return true if we just completed rack init, false otherwise
+    pub fn on_ack(&mut self, from: Baseboard) -> bool {
+        if let RackInitState::Running {
+            total_members,
+            acks,
+            unacked_envelopes,
+            ..
+        } = self
+        {
+            unacked_envelopes.retain(|envelope| envelope.to != from);
+            acks.insert(from);
+            if acks.len() == *total_members {
+                *self = RackInitState::Initialized;
+                return true;
+            }
+        }
+        false
+    }
+    pub fn timer_expired(&self, now: Ticks, timeout: Ticks) -> bool {
+        if let RackInitState::Running { start, .. } = self {
+            now.saturating_sub(*start) > timeout
+        } else {
+            false
+        }
+    }
+
+    pub fn is_initialized(&self) -> bool {
+        matches!(self, RackInitState::Initialized)
+    }
+
+    pub fn unacked_peers(&self) -> BTreeSet<Baseboard> {
+        if let RackInitState::Running { unacked_envelopes, .. } = self {
+            unacked_envelopes
+                .iter()
+                .map(|envelope| envelope.to.clone())
+                .collect()
+        } else {
+            BTreeSet::new()
+        }
+    }
+}
+
+/// State stored on the M.2s.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct PersistentState {
+    // The generation number for ledger writing purposes on both M.2s
+    pub version: u32,
+    pub fsm_state: PersistentFsmState,
+}
+
+// Based on the state machine of how `PersistentState` gets updated, we can
+// deterministically infer a version.
+impl From<PersistentFsmState> for PersistentState {
+    fn from(fsm_state: PersistentFsmState) -> Self {
+        use PersistentFsmState::*;
+        let version = match &fsm_state {
+            Uninitialized => 0,
+            InitialMember { distributed_shares, .. } => {
+                (1 + distributed_shares.len()).try_into().unwrap()
+            }
+            Learning => 1,
+            Learned { .. } => 2,
+        };
+        PersistentState { version, fsm_state }
+    }
+}
+
+/// A subset of `State` that must be persisted to the M.2s
+#[derive(Debug, Serialize, Deserialize)]
+pub enum PersistentFsmState {
+    Uninitialized,
+    InitialMember {
+        pkg: SharePkgV0,
+        distributed_shares: BTreeMap<Baseboard, ShareIdx>,
+    },
+    Learning,
+    Learned {
+        pkg: LearnedSharePkgV0,
+    },
+}
+
+impl From<State> for PersistentFsmState {
+    fn from(value: State) -> Self {
+        match value {
+            State::Uninitialized(_) => Self::Uninitialized,
+            State::InitialMember(InitialMemberState {
+                pkg,
+                distributed_shares,
+                ..
+            }) => Self::InitialMember { pkg, distributed_shares },
+            State::Learning(_) => Self::Learning,
+            State::Learned(LearnedState { pkg, .. }) => Self::Learned { pkg },
```

Comment:
This `pkg` also contains an `encrypted_shares` member variable.  Is it always guaranteed to be empty for late arrivals?  If so, should that be enforced at the type level?

Reply:
outdated